bwscanner/attacher.py

```diff
@@ -118,47 +108,80 @@ def stream_closed(self, *args, **kw):
-        d2.addCallback(lambda port: config.__setattr__('ControlPort', port))
-        return d2
+    if "UseMicroDescriptors" in new_options:
+        if tor_config.UseMicroDescriptors != new_options["UseMicroDescriptors"]:
+            log.debug("Changing UseMicroDescriptors from {current} to {new}.",
+                      current=tor_config.UseMicroDescriptors,
+                      new=new_options["UseMicroDescriptors"])
+            return True
+    return False
+
 
-    def launch_and_get_state(ignore):
-        d2 = launch_tor(config, reactor, stdout=sys.stdout)
-        d2.addCallback(lambda tpp: TorState(tpp.tor_protocol).post_bootstrap)
-        return d2
-    return get_random_tor_ports().addCallback(launch_and_get_state)
+def wait_for_newconsensus(tor_state):
+    got_consensus = defer.Deferred()
 
+    def got_newconsensus(event):
+        log.debug("Got NEWCONSENSUS event: {event}", event=event)
+        got_consensus.callback(event)
+        tor_state.protocol.remove_event_listener('NEWCONSENSUS', got_newconsensus)
 
-def update_tor_config(tor, config):
+    tor_state.protocol.add_event_listener('NEWCONSENSUS', got_newconsensus)
+    return got_consensus
+
+
+@defer.inlineCallbacks
+def connect_to_tor(launch_tor, circuit_build_timeout, control_port=None,
+                   tor_overrides=None):
     """
-    Update the Tor config from a dict of config key: value pairs.
+    Launch or connect to a Tor instance
+
+    Configure Tor with the passed options and return a Deferred
     """
-    config_pairs = [(key, value) for key, value in config.items()]
-    d = tor.protocol.set_conf(*itertools.chain.from_iterable(config_pairs))
-    return d.addCallback(lambda result: tor)
-
-
-def setconf_singleport_exit(tor):
-    port = available_tcp_port(reactor)
-
-    def add_single_port_exit(port):
-        tor.protocol.set_conf('PublishServerDescriptor', '0',
-                              'PortForwarding', '1',
-                              'AssumeReachable', '1',
-                              'ClientRejectInternalAddresses', '0',
-                              'OrPort', 'auto',
-                              'ExitPolicyRejectPrivate', '0',
-                              'ExitPolicy', 'accept 127.0.0.1:{}, reject *:*'.format(port))
-    return port.addCallback(add_single_port_exit).addCallback(
-        lambda ign: tor.routers[tor.protocol.get_info("fingerprint")])
+    # Options for spawned or running Tor to load the correct descriptors.
+    tor_options = {
+        'LearnCircuitBuildTimeout': 0,  # Disable adaptive circuit timeouts.
+        'CircuitBuildTimeout': circuit_build_timeout,
+        'UseEntryGuards': 0,  # Disable UseEntryGuards to avoid PathBias warnings.
+        'UseMicroDescriptors': 0,
+        'FetchUselessDescriptors': 1,
+        'FetchDirInfoEarly': 1,
+        'FetchDirInfoExtraEarly': 1,
+    }
+
+    if tor_overrides:
+        tor_options.update(tor_overrides)
+
+    if launch_tor:
```

Comment:
ideally we should enable the unit test that tests launch_tor option as well. I think it's disabled currently.